% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/SAIGE_SKAT_NULL_Model_usingSKATLib_p.R
\name{SKATtest_usingSKATLib}
\alias{SKATtest_usingSKATLib}
\title{Run single variant score tests with SPA based on the logistic mixed model.}
\usage{
SKATtest_usingSKATLib(
  dosageFile = "",
  dosageFileNrowSkip = 0,
  dosageFileNcolSkip = 0,
  dosageFilecolnamesSkip = c("SNPID", "CHR", "POS", "Allele0", "Allele1"),
  dosageFileChrCol = "CHR",
  bgenFile = "",
  bgenFileIndex = "",
  vcfFile = "",
  vcfFileIndex = "",
  vcfField = "DS",
  savFile = "",
  savFileIndex = "",
  sampleFile = "",
  idstoExcludeFile = "",
  idstoIncludeFile = "",
  rangestoExcludeFile = "",
  rangestoIncludeFile = "",
  chrom = "",
  start = 1,
  end = 2.5e+08,
  minMAC = 0.5,
  minMAF = 0,
  maxMAFforGroupTest = 1,
  minInfo = 0,
  SKATmodelFile = "",
  SAIGEOutputFile = "",
  numLinesOutput = 10000,
  IsOutputAFinCaseCtrl = FALSE,
  groupFile = "",
  condition = "",
  kernel = "linear.weighted",
  method = "optimal.adj",
  weights.beta = c(1, 25),
  r.corr = 0
)
}
\arguments{
\item{dosageFile}{character. Path to dosage file. Each line contains dosages for a marker to be tested}

\item{dosageFileNrowSkip}{integer(>=0). Number of lines to be skiped in the dosage file.}

\item{dosageFileNcolSkip}{integer(>=0). Number of columns to be skiped in the dosage file}

\item{dosageFilecolnamesSkip}{vector of characters. The column names of the skipped columns. Default: c("SNPID", "CHR", "POS", "Allele0", "Allele1")}

\item{dosageFileChrCol}{string. The column name for the chromosome column. Must be in the dosageFilecolnamesSkip. Required If LOCO = TRUE and chrom ="".}

\item{bgenFile}{character. Path to bgen file. Currently version 1.2 with 8 bit compression is supported}

\item{bgenFileIndex}{character. Path to the .bgi file (index of the bgen file)}

\item{vcfFile}{character. Path to vcf file}

\item{vcfFileIndex}{character. Path to index for vcf file by tabix, ".tbi" by "tabix -p vcf file.vcf.gz"}

\item{vcfField}{character. genotype field in vcf file to use. "DS" for dosages or "GT" for genotypes. By default, "DS".}

\item{savFile}{character. Path to sav file}

\item{savFileIndex}{character. Path to index for sav file .s1r}

\item{sampleFile}{character. Path to the file that contains one column for IDs of samples in the dosage, vcf, sav, or bgen file with NO header}

\item{idstoExcludeFile}{character. Path to the file containing variant ids to be excluded from the bgen or vcf file}

\item{idstoIncludeFile}{character. Path to the file containing variant ids to be included from the bgen or vcf file}

\item{rangestoExcludeFile}{character. Path to the file containing genome regions to be excluded from the bgen file. The file contains three columns for chromosome, start, and end respectively with no header}

\item{rangestoIncludeFile}{character. Path to the file containing genome regions to be included from the bgen file. The file contains three columns for chromosome, start, and end respectively with no header}

\item{chrom}{character. string for the chromosome to include from vcf file. Required for vcf file. If LOCO is specified, providing chrom will save computation cost}

\item{start}{numeric. start genome position to include from vcf file.}

\item{end}{numeric. end genome position to include from vcf file.}

\item{minMAC}{numeric. Minimum minor allele count of markers to test. By default, 1. The higher threshold between minMAC and minMAF will be used}

\item{minMAF}{numeric. Minimum minor allele frequency of markers to test. By default 0. The higher threshold between minMAC and minMAF will be used}

\item{maxMAFforGroupTest}{numeric. Maximum minor allele frequency of markers to test in group test. By default 1.}

\item{minInfo}{numeric. Minimum imputation info of markers to test (in bgen file)}

\item{SAIGEOutputFile}{character. Path to the output file containing the SPAGMMAT test results}

\item{numLinesOutput}{numeric. Output results for how many marker each time.}

\item{IsOutputAFinCaseCtrl}{logical. Whether to output allele frequency in cases and controls. By default, FALSE}

\item{groupFile}{character. Path to the group file containing one column "GeneID", and one column for ID of the tested genetic markers delimited by comma. This file is space-delimited can only work with the bgen,sav,and vcf format}

\item{GMMATmodelFile}{character. Path to the input file containing the glmm model, which is output from previous step. Will be used by load()}

\item{varianceRatioFile}{character. Path to the input file containing the variance ratio, which is output from the previous step}

\item{Cutoff}{by default = 2 (SPA test would be used when p value < 0.05 under the normal approximation)}

\item{IsSparse}{logical. Whether to exploit the sparsity of the genotype vector for less frequent variants to speed up the SPA tests or not for dichotomous traits. By default, TRUE}

\item{condition.}{For conditional analysis. Genetic marker ids (chr:pos_ref/alt) seperated by comma. e.g.chr3:101651171_C/T,chr3:101651186_G/A. Note that currently conditional analysis is only for vcf/sav input.}
}
\value{
SAIGEOutputFile
}
\description{
Run single variant score tests with SPA based on the logistic mixed model.
}
